onnxscript/function_libs/torch_aten/ops/core.py

```diff
@@ -269,44 +269,80 @@ def aten_any(self: TensorType) -> TensorType:
     raise NotImplementedError()
 
 
-@torch_op("aten::arange")
+def _range_supported(dtype: int) -> bool:
+    """Returns true if the dtype is supported by the ONNX Range op."""
+    return dtype in {
+        DOUBLE.dtype,
+        FLOAT.dtype,
+        INT16.dtype,
+        INT32.dtype,
+        INT64.dtype,
+    }
+
+
+@torch_op("aten::arange", trace_only=True)
 def aten_arange(end: Union[DOUBLE, FLOAT, INT16, INT32, INT64], dtype: int = -1) -> TensorType:
     # arange(Scalar end, *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None) -> Tensor
 
-    # Cast input to double if dtype is specified, because the input dtype may be e.g. bool
-    # which Range does not support. The output type is ensured because the output
-    # is casted to the specified dtype.
-    if dtype != -1:
-        end = op.Cast(end, to=DOUBLE.dtype)
+    # NOTE: trace_only because both if branches need to be the same type, but we have
+    # a cast in the if branch.
 
-    result = op.Range(0, end, 1)
-    if dtype != -1:
-        result = op.Cast(result, to=dtype)
+    if _range_supported(dtype):
+        end = op.Cast(end, to=dtype)
+        zero = op.Cast(0, to=dtype)
+        one = op.Cast(1, to=dtype)
+        result = op.Range(zero, end, one)
+    else:
+        # Cast input to float if dtype is not supported by Range,
+        # because the input dtype may be e.g. bool
+        # which Range does not support. The output type is ensured because the output
+        # is casted to the specified dtype.
+        end_ = end
+        end = op.Cast(end, to=FLOAT.dtype)
+        zero = op.Constant(value_float=0.0)
+        one = op.Constant(value_float=1.0)
+        result = op.Range(zero, end, one)
+        if dtype != -1:
+            result = op.Cast(result, to=dtype)
+        else:
+            result = op.CastLike(result, end_)
 
     return result
 
 
-@torch_op("aten::arange", overload=True)
+@torch_op("aten::arange", overload=True, trace_only=True)
```

Comment:
(Not part of this PR, but wondering): What does the name `TRealUnlessFloat16OrInt8` mean? The name seems unconnected to the actual definition.

Reply:
`DOUBLE, FLOAT, INT16, INT32, INT64`. I think it's accurate? Maybe I am missing something?

Reply:
Ok. There's bfloat16 missing in the name. And the word `Unless` is misleading. Perhaps it is supposed to be `Less` or `Minus` or `ButNot` or `AndNot`.

Reply:
How about `TRealNotFloat16OrBFloat16OrInt8`

Reply:
Created https://github.com/microsoft/onnx-script/issues/425 to track